xds/src/main/java/io/grpc/xds/LookasideLb.java

```diff
@@ -49,169 +48,223 @@
+          new ErrorPicker(
+              Status.UNAVAILABLE.withDescription("ATTR_LOAD_BALANCING_CONFIG not available")));
+      return;
+    }
     ConfigOrError cfg =
         XdsLoadBalancerProvider.parseLoadBalancingConfigPolicy(newRawLbConfig, lbRegistry);
     if (cfg.getError() != null) {
-      throw cfg.getError().asRuntimeException();
+      // This will not happen when the service config error handling is implemented.
+      // For now simply go to TRANSIENT_FAILURE.
+      lookasideLbHelper.updateBalancingState(TRANSIENT_FAILURE, new ErrorPicker(cfg.getError()));
+      return;
     }
-    XdsConfig xdsConfig = (XdsConfig) cfg.getConfig();
-
-    final String newBalancerName = xdsConfig.balancerName;
-
-    // The is to handle the legacy usecase that requires balancerName from xds config.
-    if (!newBalancerName.equals(balancerName)) {
-      balancerName = newBalancerName; // cache the name and check next time for optimization
-      Node nodeFromResolvedAddresses = resolvedAddresses.getAttributes().get(XDS_NODE);
-      final Node node;
-      if (nodeFromResolvedAddresses == null) {
-        node = Node.newBuilder()
-            .setMetadata(Struct.newBuilder()
-                .putFields(
-                    "endpoints_required",
-                    Value.newBuilder().setBoolValue(true).build()))
-            .build();
+    XdsConfig newXdsConfig = (XdsConfig) cfg.getConfig();
+    ObjectPool<XdsClient> xdsClientRefFromResolver = attributes.get(XdsAttributes.XDS_CLIENT_REF);
+    ObjectPool<XdsClient> xdsClientRef;
+
+    // Init XdsClient.
+    if (xdsClient == null) {
+      // There are three usecases:
+      // 1. The EDS-only legacy usecase that requires balancerName from xds config.
+      //    Note: we don't support balancerName change.
+      //    TODO(zdapeng): Remove the legacy case.
+      // 2. The EDS-only with bootstrap usecase:
+      //    The name resolver resolves a ResolvedAddresses with an XdsConfig without balancerName
+      //    field. Use the bootstrap information to create a channel.
+      // 3. Non EDS-only usecase:
+      //    XDS_CLIENT_REF attribute is available from ResolvedAddresses either from
+      //    XdsNameResolver or CDS policy.
+      //
+      // We assume XdsConfig switching happens only within one usecase, and there is no switching
+      // between different usecases.
+      if (newXdsConfig.balancerName != null) {
+        // This is the EDS-only legacy usecase that requires balancerName from xds config.
+        channel = initLbChannel(
+            lookasideLbHelper, newXdsConfig.balancerName, Collections.<ChannelCreds>emptyList());
+        xdsClientRef = new RefCountedXdsClientObjectPool(new XdsClientFactory() {
+          @Override
+          XdsClient createXdsClient() {
+            return new XdsComms2(
+                channel, lookasideLbHelper, new ExponentialBackoffPolicy.Provider(),
+                GrpcUtil.STOPWATCH_SUPPLIER, Node.getDefaultInstance());
```

Comment:
Why are you passing `Node.getDefaultInstance()` when you can pass the actual value `bootstrapper.readBootstrap().getNode()`?